### PR TITLE
[listennotes.com] Add extractor

### LINKED ISSUE #5262: support downloading from listennotes.com
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

Europe

### Example URLs

https://www.listennotes.com/podcasts/thriving-on-overload/tim-oreilly-on-noticing-KrDgvNb_u1n/
https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/

### Provide a description that is worded well enough to be understood

Listen Notes is a website which aggregates podcasts from all over the internet, into one searchable platform.
This is a great place to find podcasts based on search criteria.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '-o', '/mnt/media/video/17-%(title).200s-%(upload_date>%Y-%m-%d)s.%(ext)s', 'https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/']
[debug] User config "/home/karel/.config/yt-dlp/config": ['--restrict-filenames', '--no-check-certificate', '--sub-lang', 'en', '-f', '22/bestvideo[ext=mp4][width<=1920]+bestaudio[ext=m4a]/mp4/m4a/mp3']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out UTF-8, error UTF-8, screen UTF-8
[debug] yt-dlp version 2022.10.04 [4e0511f27] (debian*) API
[debug] Python 3.7.3 (CPython 64bit) - Linux-5.10.109-grsec-x86_64-with-debian-10.13 (glibc 2.28)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.1.9-0, ffprobe 4.1.9-0
[debug] Optional libraries: Cryptodome-3.6.1, brotli-1.0.7, certifi-2018.08.24, mutagen-1.40.0, secretstorage-2.3.1, sqlite3-2.6.0, websockets-7.0
[debug] Proxy map: {}
[debug] Loaded 1690 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.10.04, Current version: 2022.10.04
yt-dlp is up to date (2022.10.04)
[debug] [generic] Extracting URL: https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/
[generic] episode-177-wireguard-with-lwEA3154JzG: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] episode-177-wireguard-with-lwEA3154JzG: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
[debug] Identified a twitter:player iframe
[debug] [generic] Extracting URL: https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/embed/
[generic] embed: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] embed: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
ERROR: Unsupported URL: https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/embed/
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/yt_dlp/YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/yt_dlp/YoutubeDL.py", line 1553, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3/dist-packages/yt_dlp/extractor/common.py", line 672, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3/dist-packages/yt_dlp/extractor/generic.py", line 3062, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.listennotes.com/podcasts/ask-noah-show/episode-177-wireguard-with-lwEA3154JzG/embed/
```

### PR DESCRIPTION
**Pull request by lksj**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Add extractor for listennotes.com

Fixes #5262

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/listennotes.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ class ListenNotesIE(InfoExtractor):
     }]
 
     def _real_extract(self, url):
-        id = self._match_id(url).replace('/', '').rsplit('-', 1)[1]
+        id = self._match_id(url).rsplit('-', 1)[1]
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:28]: Why not change the `VALID_URL` to get right id ?. I recommend to use other variable name other than `id`

File: yt_dlp/extractor/listennotes.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            'ext': 'mp3',
+            'title': 'Episode 177: WireGuard with Jason Donenfeld',
+            'description': '01:04:21 - Jason Donenfeld lead developer joins us this hour to discuss WireGuard, an extremely simple yet fast and modern VPN that utilizes state-of-the-art c…'
+        }
+    }]
+
+    def _real_extract(self, url):
+        id = self._match_id(url).rsplit('-', 1)[1]
+        webpage = self._download_webpage(url, id)
+        data_json = self._parse_json(self._search_regex(r'<script id="original-content" type="application/json">\s*({"uuid".+})', webpage, 'content'), id)
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:30]: i think `search_json` can be used here

File: yt_dlp/extractor/listennotes.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        audio_url = data_json['audio']
+        description = re.sub(r"""\s{2,}""", " ", self._og_search_description(webpage))
+        title = self._html_search_regex(r'<h1\s*class=".+?">\s*<a\s*href=".+?"\s*title="(.+?)"\s*class=".+?">\s*.+\s*</a>\s*</h1>', webpage, 'title')
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:33]: inline in return

**Review by HobbyistDev — COMMENTED** [review]
I think you can set fallback (eg: `og:title` for `title`, etc) and maybe you can extract more field in `json_ld`

File: yt_dlp/extractor/listennotes.py (hunk 1 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py]: This is done in e7d09bef4197cbb6d20b7e397b98e3cbce57f0fb

File: yt_dlp/extractor/listennotes.py (hunk 3 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py]: Refactored in 0f7a87cafd1b6d7885ae936bdddf2d933bbe401c

File: yt_dlp/extractor/listennotes.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            'ext': 'mp3',
+            'title': 'Episode 177: WireGuard with Jason Donenfeld - Ask Noah Show (podcast)',
+            'description': '01:04:21 - Jason Donenfeld lead developer joins us this hour to discuss WireGuard, an extremely simple yet fast and modern VPN that utilizes state-of-the-art c…'
+        }
+    }]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+        webpage = self._download_webpage(url, audio_id)
+        title = self._html_search_regex(r'<h1\s*class=".+?">\s*<a\s*href=".+?"\s*title="(.+?)"\s*class=".+?">\s*.+\s*</a>\s*</h1>', webpage, 'title')
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:30]: Why define title twice ?

File: yt_dlp/extractor/listennotes.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        audio_url = self._search_json(
+            r'<script id="original-content" type="application/json">\s*', webpage, 'content', audio_id).get('audio')
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:32]: I think you can set like this
```suggestion
        json_data = self._search_json(
            r'<script id="original-content" type="application/json">\s*', webpage, 'content', audio_id)
```

and then return audio in return like this:
```
     'url': json_data['audio']
```

File: yt_dlp/extractor/listennotes.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        description = self._html_search_meta(
+            ['og:description', 'description', 'twitter:description'], webpage, 'description', default=None)
+        if description is not None:
+            description = re.sub(r'\s{2,}', ' ', description)
```
Comment (HobbyistDev) [inline yt_dlp/extractor/listennotes.py:36]: as far i can see, we can set the `duration` and `description` from the `description` field in `json_data`. It looks like this:
```suggestion
        description_raw = self._html_search_meta(
            ['og:description', 'description', 'twitter:description'], webpage, 'description')
        duration, description = self._search_regex(
             r'(?P<duration>\d+)\s*-\s*(?P<description>[\w]+)', description_raw, 'description').group('duration', 'description')
```
However, i didn't test this. Please correct me if i'm wrong

**Review by HobbyistDev — COMMENTED** [review]
i suggest you to take extract `json ld` as well using `self._search_json_ld`, like dirkf said

**Comment by HobbyistDev** [comment]
Please avoid force-pushing

File: yt_dlp/extractor/listennotes.py (hunk 4 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py:30]: I don't understand. Where is double definition?

**Comment by lksj** [comment]
> i suggest you to take extract `json ld` as well using `self._search_json_ld`, like dirkf said

i tried to extract it. Nothing interesting was found.

File: yt_dlp/extractor/listennotes.py (hunk 6 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py:36]: This code will not work in some cases.

1. Description field can contain a picture, which has many white-spaces in the html code. They need to be cleaned up.
2 Description field may be None.

Why need extracting duration field? Where is it used?

File: yt_dlp/extractor/listennotes.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/listennotes.py:30]: L37

**Comment by lksj** [comment]
> Please avoid force-pushing

OK. Do I need to squash many commits to one?

**Comment by pukkandan** [comment]
> Do I need to squash many commits to one?

No. I will squash during merge. Feel free to make as many commits as you need

File: yt_dlp/extractor/listennotes.py (hunk 5 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/listennotes.py:32]: This is true for non-fatal field. I think the `url` field should fatal.

File: yt_dlp/extractor/listennotes.py (hunk 4 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py:30]: Oh, i'm sorry. I forgote delete L30. I will fix it.

**Comment by HobbyistDev** [comment]
> Why need extracting duration field? Where is it used?

Some people needs video metadata for various reason (eg: analytics, etc). I think it's better if we can get all metadata that we can get.
 
For information what field we can used, please take a look at `extractor/common.py` especially at `InfoExtractor` class

File: yt_dlp/extractor/listennotes.py (hunk 4 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py]: Fixed. 33bd35dc345c3093da73fbea13bac5a41ebc7e9d

File: yt_dlp/extractor/listennotes.py (hunk 6 above)
Comment (lksj) [reply yt_dlp/extractor/listennotes.py]: Done. b103f1eda42e4746e5fcecce8dc9d14b3b6d23c6

**Comment by pukkandan** [comment]
I added fallback, full description and more fields. Pls have a look and let me know if you think the fields I chose make sense

**Pull request merged**